### ISSUE TITLE
AIX: - fs_event_watch_file_current_dir - Intermittent?

### ISSUE THREAD

**Issue by mhdawson** [libuv/libuv#2118]
* **Version**: 1.x
* **Platform**: master

```bash
not ok 60 - fs_event_watch_file_current_dir
# timeout
# Output from process `fs_event_watch_file_current_dir`: (no output)
```

**Comment by mhdawson** [comment]
@gireeshpunathil could you take a look.  Seems to fail in every run now.

**Comment by gireeshpunathil** [comment]
unable to reproduce the failure in several runs (running in CI box itself). @mhdawson - do you have a CI link where this test shows as failing?

**Comment by mhdawson** [comment]
I see the second one I started passed.  I saw the intervening failing run an believed it was a consistent failure.  Obvious I was wrong on that.  May be intermittent or a one off. 

I'd suggest we just run the single test itself say 1000 times and if it does not recreate, close.

**Comment by gireeshpunathil** [comment]
I am able to recreate the failure. It is basically a hang, that the test harness kills after the timeout.

**Comment by gireeshpunathil** [comment]
a simple file system watcher test that installs a listener on file modification, then opens the file and writes a word. The listener never gets fired.

relevant sections from passing case:
```strace
5308492: 58261669: mkdir("/aha/fs/modFile.monFactory", 0775) Err#17 EEXIST
5308492: 58261669: mkdir("/aha/fs/modFile.monFactory/home", 0775) Err#17 EEXIST
5308492: 58261669: mkdir("/aha/fs/modFile.monFactory/home/iojs", 0775) Err#17 EEXIST
5308492: 58261669: mkdir("/aha/fs/modFile.monFactory/home/iojs/gireesh", 0775) Err#17 EEXIST
5308492: 58261669: mkdir("/aha/fs/modFile.monFactory/home/iojs/gireesh/libuv", 0775) Err#17 EEXIST
5308492: 58261669: kopen("/aha/fs/modFile.monFactory/home/iojs/gireesh/libuv/watch_file.mon", O_RDWR|O_CREAT|O_LARGEFILE, S_ISUID|S_IRUSR|S_IXUSR|S_IROTH) = 10
5308492: 58261669: kwrite(10, " C H A N G E D = Y E S ;".., 48) = 48
5308492: 58261669: pollset_ctl(1, 0x2FF22978, 1)                = 0
5308492: 58261669: pollset_ctl(1, 0x2FF22978, 1)                = 0
5308492: 58261669: pollset_ctl(1, 0x2FF22978, 1)                = 0
5308492: 58261669: _pollset_poll(1, 0x2FF20910)         = 0
5308492: 58261669: kopen("watch_file", O_RDWR|O_LARGEFILE) = 11
5308492: 58261669: kfcntl(11, F_SETFD, 0x00000001)      = 0
5308492: 58261669: kwrite(11, " f o o\0", 4)            = 4
5308492: 58261669: close(11)                            = 0
5308492: 58261669: _pollset_poll(1, 0x2FF20910)         = 1
```

failing case:

```strace
52822164: 77463725: mkdir("/aha", 0775)                 Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs", 0775)              Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs/modFile.monFactory", 0775) Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs/modFile.monFactory/home", 0775) Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs/modFile.monFactory/home/iojs", 0775) Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs/modFile.monFactory/home/iojs/gireesh", 0775) Err#17 EEXIST
52822164: 77463725: mkdir("/aha/fs/modFile.monFactory/home/iojs/gireesh/libuv", 0775) Err#17 EEXIST
52822164: 77463725: kopen("/aha/fs/modFile.monFactory/home/iojs/gireesh/libuv/watch_file.mon", O_RDWR|O_CREAT|O_LARGEFILE, S_ISUID|S_IRUSR|S_IXUSR|S_IROTH) = 10
52822164: 77463725: kwrite(10, " C H A N G E D = Y E S ;".., 48) = 48
52822164: 77463725: kopen("watch_file", O_RDWR|O_LARGEFILE) = 11
52822164: 77463725: kfcntl(11, F_SETFD, 0x00000001)     = 0
52822164: 77463725: kwrite(11, " f o o\0", 4)           = 4
52822164: 77463725: close(11)                           = 0
52822164: 77463725: pollset_ctl(19, 0x2FF22978, 1)      = 0
52822164: 77463725: pollset_ctl(19, 0x2FF22978, 1)      = 0
52822164: 77463725: pollset_ctl(19, 0x2FF22978, 1)      = 0
52822164: 77463725: _pollset_poll(19, 0x2FF20910)               = 0
52822164: 77463725: _pollset_poll(19, 0x2FF20910) (sleeping...)
```
clearly, the difference is in terms of the `pollset_ctl` calls:  in passing case the event is registered before the file write, whereas in the failing case, after! how could that be?

**Comment by gireeshpunathil** [comment]
```c
  uv_loop_t* loop;
  int r;
  uv_buf_t buf;
  loop = uv_default_loop();
  r = uv_fs_open(NULL, &req, name, O_WRONLY | O_CREAT, S_IWUSR | S_IRUSR, NULL);
  r = uv_fs_close(NULL, &req, file, NULL);
  r = uv_fs_event_init(loop, &fs_event);
  r = uv_fs_event_start(&fs_event, fs_event_cb_file_current_dir, "watch_file", 0);
  r = uv_fs_open(NULL, &req, "watch_file", O_RDWR, 0, NULL);
  buf = uv_buf_init("foo", 4);
  r = uv_fs_write(NULL, &req, file, &buf, 1, -1, NULL);
  r = uv_fs_close(NULL, &req, file, NULL);
  uv_run(loop, UV_RUN_DEFAULT);
```

```js
const fs = require('fs')
fs.watch('./foo.txt', (t, n) => {
  console.log(`${n} got ${t}'d!`)
})
fs.writeFileSync('./foo.txt', 'foo')
```

The hang is recreatable in C or JS as above, AIX. All other platforms report file change event. Not a regression, issue exist right from inception of `watch` feature in AIX.

The fundamental issue here is that the file modification is performed in the same loop itreation in which the event monitors are created but not yet registered, instead only enqueued into the poll structure, that will be taken up with the OS on the loop edge only.

AHAFS (that provides the file watch facility for AIX) specifies that "monitoring of events does not begin until the monitoring application issues a select() or a blocking read() call"

https://www.ibm.com/developerworks/aix/library/au-aix_event_infrastructure/ (waiting on event occurrences section)

So basically we missed the event, as we did not register for it in time.

I don't think it is solvable with the current libuv design. At the same time, I also think this is not a major concern, as the event monitors and event sources may not be in the same process / same control flow sequence in the most common use cases.

Would like to hear from others.

[ what I don't really understand is why the `timer` event was fired in the same loop in the original test case occasionally, to make it flaky ]

**Comment by gireeshpunathil** [comment]
thanks @bnoordhuis - that works!

with this,

```patch
diff --git a/deps/uv/src/unix/aix.c b/deps/uv/src/unix/aix.c
index 44c9cf5b63..42bbf8db82 100644
--- a/deps/uv/src/unix/aix.c
+++ b/deps/uv/src/unix/aix.c
@@ -768,6 +768,11 @@ int uv_fs_event_start(uv_fs_event_t* handle,
 
   uv__io_start(handle->loop, &handle->event_watcher, POLLIN);
 
+  struct timeval t;
+  t.tv_sec = 0;
+  t.tv_usec = 0;
+  select(1, &fd, NULL, NULL, &t);
+
   return 0;
 #else
   return UV_ENOSYS;
```

this is the pattern I observe, and it did not hang in 2K iterations!

```strace
6225980: 80871583: kopen("/aha/fs/modFile.monFactory/home/iojs/gireesh/libuv/watch_file.mon", O_RDWR|O_CREAT|O_LARGEFILE, S_ISUID|S_IRUSR|S_IXUSR|S_IRGRP|S_IROTH) = 9
6225980: 80871583: kwrite(9, " C H A N G E D = Y E S ;".., 48) = 48
6225980: 80871583: _select(1, 0x2FF22928, 0x00000000, 0x00000000, 0x2FF20928) = 0
6225980: 80871583: pollset_ctl(0, 0x2FF22998, 1)                = 0
6225980: 80871583: pollset_ctl(0, 0x2FF22998, 1)                = 0
6225980: 80871583: pollset_ctl(0, 0x2FF22998, 1)                = 0
6225980: 80871583: _pollset_poll(0, 0x2FF20930)         = 0
6225980: 80871583: kopen("watch_file", O_RDWR|O_LARGEFILE) = 10
6225980: 80871583: kfcntl(10, F_SETFD, 0x00000001)      = 0
6225980: 80871583: kwrite(10, " f o o\0", 4)            = 4
6225980: 80871583: close(10)                            = 0
6225980: 80871583: _pollset_poll(0, 0x2FF20930)         = 1
```

are you going for the PR?

**Comment by bnoordhuis** [comment]
You're welcome to open one, you wrote it and I can't test locally anyway. :-)
